storage/zynq7000-flash/qspi.c

```diff
@@ -0,0 +1,436 @@
+/*
+ * Phoenix-RTOS
+ *
+ * Quad-SPI Controller driver
+ *
+ * Copyright 2021, 2022 Phoenix Systems
+ * Author: Hubert Buczynski
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include "qspi.h"
+
+#include <errno.h>
+#include <string.h>
+#include <sys/time.h>
+#include <sys/mman.h>
+#include <sys/threads.h>
+#include <sys/platform.h>
+#include <sys/interrupt.h>
+
+#include <phoenix/arch/zynq7000.h>
+
+
+enum { cr = 0, sr, ier, idr, imr, er, dr, txd00, rxd, sicr, txth, rxth, gpio,
+	   lpbk = 0xe, txd01 = 0x20, txd10, txd11,
+	   lqspi_cr = 0x28, lqspi_sr, modid = 0x3f };
+
+
+struct {
+	volatile uint32_t *base;
+	handle_t cond;
+	handle_t inth;
+	handle_t lock;
+} qspi_common;
+
+
+static inline void qspi_dataMemoryBarrier(void)
+{
+	__asm__ volatile ("dmb");
+}
+
+
+static int qspi_irqHandler(unsigned int n, void *arg)
+{
+	*(qspi_common.base + idr) |= (1 << 2);
+
+	return 1;
+}
+
+
+static inline time_t qspi_timeMsGet(void)
```

Comment:
Code duplication - flashdrv_timeMsGet

Reply:
Yeah, there is code duplication but adding this function to `qspi` external interface would looke strange, creating some common part would also not be a solution. So that I decided to implement it in two places.